### DISCUSSION TITLE
Non uniform mass distribution for RigidBody3D

### DISCUSSION THREAD

**Discussion by perococco** [godotengine/godot-proposals#7600] [3D]
Motivation
-----------
I want to develop a space simulation game (something like KSP) in which you could create rockets, spaceships by assembling different parts (engine, tank, cockpit). Each part has its own mass that might change over time (mostly for tanks).

In Godot, I use a [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html) for the spaceship and for the parts a pair of ([MeshInstance3D](https://docs.godotengine.org/en/stable/classes/class_meshinstance3d.html),[CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html)) that you can choose from a list of parts.

As far as I can tell (I gather these information by browsing the source code) the [CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html) is used to detect collisions (obviously) and to compute the [mass_properties](https://github.com/godotengine/godot/blob/fa3428ff25bc577d2a3433090478a6d615567056/servers/physics_3d/godot_body_3d.cpp#L55C26-L55C26) (i.e. the center of mass and the inertia tensor of the body).

The issue I face, is that the computation of the mass properties assumes that the mass is uniformly distributed over the [CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html) in the [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html) (the effect of a [CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html) during the computation is based on its volume only).

I would like to be able to define a non uniformly distributed mass over the [CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html) in a given [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html).

Implications
-----------
This implies a modification on how the mass of a [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html) is defined and how the mass properties are computed. 

1. Add a _Mass Mode_ with 2 choices for the [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html)
   1. `Uniform` (the default choice)
   2. `Non Uniform`

   In `Uniform` mode, we have the current behavior of Godot and the field _Mass_ is visible and editable for the [RigidBody3D](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html) in the editor. In `Non Uniform` mode the _Mass_ is not editable (and "not visible" or "visible and filled automatically with the computed mass" ? )

2. On each collision shape, add a field _Mass_ only visible and editable if the mass mode is `Non Uniform`.

3. On the core engine, the computation of the [mass_properties](https://github.com/godotengine/godot/blob/fa3428ff25bc577d2a3433090478a6d615567056/servers/physics_3d/godot_body_3d.cpp#L55C26-L55C26) must be updated in the case of the `Non Uniform` mode.
   1. The total mass must be computed
   2. The computation of center of mass and the inertia tensor must use the mass of the [CollisionShape3D](https://docs.godotengine.org/en/stable/classes/class_collisionshape3d.html) instead of its volume.

Although, the starting point is a specific need for my game, I feel like this a more general request to improve the 3D physic in Godot with not that much effort.

If this discussion is successful and requires development, I am ready to ensure it.

**Comment by theraot** [comment]
Be aware that you can already set the `center_of_mass` of the `RigidBody3D` (set `center_of_mass_mode` to `CENTER_OF_MASS_MODE_CUSTOM`), which is a way to represent a not uniform mass distribution. The `center_of_mass` is used for the torque component of off-center impulses, among other things. So, I believe that setting the `center_of_mass`  would be sufficient for the feature you want.

**Reply by perococco** [reply]
Yes the center of mass can be set manually, but not the inertia tensor : only a fraction of the tensor can be set (i.e. the principal components which are relative to the local frame) so no asymmetric mass distribution can be represented with the custom values for the inertia tensor (well you could but that would imply a lot of side effects to deal with in your program).

**Comment by AThousandShips** [comment]
* Duplicate of: https://github.com/godotengine/godot-proposals/issues/7033

Thank you for opening, consolidating there

**Reply by perococco** [reply]
Hi @AThousandShips, not sure what is the process here. Should I consider this feature will be implemented indirectly by with #7033, or should I comment the issue #7033 with my suggestions event though this is not exactly the same feature ?

**Reply by AThousandShips** [reply]
Add the details I'd say, what more would be required than inertia and the center of mass? The physics engine wouldn't simulate any more detailed mass properties than that
